src/tests/suite/daos_pool.c

```diff
@@ -797,20 +798,24 @@ pool_op_retry(void **state)
 	assert_rc_equal(rc, -DER_MISC);
 	print_message("success\n");
 
-	/* pool update ACL entry failure committed, "lost" reply - duplicate RPC retry */
-	test_set_engine_fail_loc(arg, leader_rank, DAOS_MD_OP_FAIL_NOREPLY | DAOS_FAIL_ONCE);
+	/* pool update ACL entry failure committed, "lost" reply - duplicate RPC retry.
+	 * NB: with dmg, fail until cleared (in case control plane "retries")
+	 */
+	test_set_engine_fail_loc(arg, leader_rank, DAOS_MD_OP_FAIL_NOREPLY | DAOS_FAIL_ALWAYS);
```

Comment:
[Question] Is it correct that this fault injection causes a related operation to complete with result `-DER_MISC` while the reply `rc` is `-DER_TIMEDOUT`? If that's correct, the test will hang at such an operation. Obviously my understanding is missing something. Could you point out what I'm missing, please? :)

Reply:
RPC handling 1: complete with -DER_MISC (and save it into svc_ops KVS) and return -DER_TIMEDOUT
RPC handling 2 (the retry): look up the previous -DER_MISC result and reply with that error rc.

If the control plane intermittently may initiate the whole thing all over again as a brand-new operation (because of its "fan out" design to multiple MS daos_server instnaces), then the sequence above should repeat.

Reply:
actually let me think about it more. I tested it and it ran successfully, but how could it be if we set fault injection for always. I'll be off for a short while and will check next week. Maybe a workaround for control plane logic is not possible with this approach.

Reply:
After reviewing the code again, and the artifacts from the latest CI testing:

1. client test_set_engine_fail_loc(fail_noreply|always)
2. gRPC 1, cart RPC handling 1 returns -DER_TIMEDOUT and saves -DER_MISC as the result of the operation
3. gRPC 1, cart RPC handling 2 sees fault injection is (still) configured, *and* also sees that this is a duplicate operation. Because it is dup, the logic inside the handler (in this case ds_pool_acl_update_handler()) replies with the retrieved status, -DER_MISC. i.e. the duplicate detection takes precedence over the fault injection configuration (which remains enabled even after the reply).

In most cases in which the control plane does not retry the gRPC from dmg utility, we're done and we have passed the test (and the test would clear the fault injection with test_set_engine_fail_loc(value=0). But let's say we hit the rare case that dmg retries. In this case steps 2 and 3 are repeated with "gRPC 2" and daos_test will ultimately see -DER_MISC returned, which is what the test expects.

I tried to find a "smoking gun" MS retry scenario among the 10 repeat loops from the Jenkins build 1 CI testing. At least for the acl update and delete handlers (which are easier to grep for than the other operation for which this applies, pool evict), I did not see the case we want to solve occur. So at this point I am going on the above code analysis that this will work when the MS on rare occasions will retry the gRPC.

Reply:
Thanks for taking a closer look at this.